{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "cell_id": "f64a7e9d-abc0-4ff1-92b6-05f5743bc6e0",
    "deepnote_cell_height": 393.0625,
    "deepnote_cell_type": "code",
    "deepnote_to_be_reexecuted": false,
    "execution_millis": 6589,
    "execution_start": 1670939976313,
    "source_hash": "24a46bf4",
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-01-10 13:34:35.147018: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "# Make numpy values easier to read.\n",
    "np.set_printoptions(precision=3, suppress=True)\n",
    "\n",
    "import tensorflow as tf\n",
    "from tensorflow.keras import layers\n",
    "from sklearn.model_selection import train_test_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "cell_id": "9bf9523d67304a9f9036fa9e85a46b89",
    "deepnote_cell_height": 649,
    "deepnote_cell_type": "code",
    "deepnote_to_be_reexecuted": false,
    "execution_millis": 97,
    "execution_start": 1670939982945,
    "source_hash": "67b0b94c",
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Solubility(%)</th>\n",
       "      <th>Yield(uM)</th>\n",
       "      <th>Yield(ug/ml)</th>\n",
       "      <th>Minus Sol</th>\n",
       "      <th>TF Sol</th>\n",
       "      <th>GroE Sol</th>\n",
       "      <th>KJE Sol</th>\n",
       "      <th>Minus(uM)</th>\n",
       "      <th>TF(uM)</th>\n",
       "      <th>...</th>\n",
       "      <th>Gene product description</th>\n",
       "      <th>Cell location</th>\n",
       "      <th>Structure (PDB) id</th>\n",
       "      <th>SCOP assignment</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Sequence length</th>\n",
       "      <th>Sequence mass</th>\n",
       "      <th>Organism</th>\n",
       "      <th>UP_ID</th>\n",
       "      <th>UP_NAME</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>thrB</td>\n",
       "      <td>32.0</td>\n",
       "      <td>2.3</td>\n",
       "      <td>78.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>homoserine kinase</td>\n",
       "      <td>Cytoplasmic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>54211; Ribosomal protein S5 domain 2-like 5506...</td>\n",
       "      <td>MVKVYAPASSANMSVGFDVLGAAVTPVDGALLGDVVTVEAAETFSL...</td>\n",
       "      <td>310</td>\n",
       "      <td>33624</td>\n",
       "      <td>Escherichia coli (strain K12 / DH10B)</td>\n",
       "      <td>B1XBC8</td>\n",
       "      <td>KHSE_ECODH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>thrC</td>\n",
       "      <td>18.0</td>\n",
       "      <td>1.7</td>\n",
       "      <td>78.0</td>\n",
       "      <td>46.53</td>\n",
       "      <td>45.80</td>\n",
       "      <td>83.44</td>\n",
       "      <td>63.61</td>\n",
       "      <td>0.863</td>\n",
       "      <td>0.761</td>\n",
       "      <td>...</td>\n",
       "      <td>threonine synthase</td>\n",
       "      <td>Cytoplasmic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>53686; Tryptophan synthase beta subunit-like P...</td>\n",
       "      <td>MKLYNLKDHNEQVSFAQAVTQGLGKNQGLFFPHDLPEFSLTEIDEM...</td>\n",
       "      <td>428</td>\n",
       "      <td>47114</td>\n",
       "      <td>Escherichia coli (strain K12)</td>\n",
       "      <td>['P00934', 'Q6LEK8']</td>\n",
       "      <td>THRC_ECOLI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>yaaX</td>\n",
       "      <td>78.0</td>\n",
       "      <td>1.2</td>\n",
       "      <td>14.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>predicted protein</td>\n",
       "      <td>Periplasmic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MKKMQSIVLALSLVLVAPMAAQAAEITLVPSVKLQIGDRDNRGYYW...</td>\n",
       "      <td>98</td>\n",
       "      <td>11355</td>\n",
       "      <td>Escherichia coli (strain K12)</td>\n",
       "      <td>P75616</td>\n",
       "      <td>YAAX_ECOLI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>yaaA</td>\n",
       "      <td>7.0</td>\n",
       "      <td>2.4</td>\n",
       "      <td>71.0</td>\n",
       "      <td>14.01</td>\n",
       "      <td>26.65</td>\n",
       "      <td>52.96</td>\n",
       "      <td>17.69</td>\n",
       "      <td>1.199</td>\n",
       "      <td>1.014</td>\n",
       "      <td>...</td>\n",
       "      <td>conserved protein</td>\n",
       "      <td>Cytoplasmic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MLILISPAKTLDYQSPLTTTRYTLPELLDNSQQLIHEARKLTPPQI...</td>\n",
       "      <td>258</td>\n",
       "      <td>29586</td>\n",
       "      <td>Escherichia coli (strain K12 / DH10B)</td>\n",
       "      <td>B1XBD1</td>\n",
       "      <td>YAAA_ECODH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>talB</td>\n",
       "      <td>85.0</td>\n",
       "      <td>2.7</td>\n",
       "      <td>94.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>transaldolase B</td>\n",
       "      <td>Cytoplasmic</td>\n",
       "      <td>1ONR</td>\n",
       "      <td>51569; Aldolase</td>\n",
       "      <td>MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...</td>\n",
       "      <td>317</td>\n",
       "      <td>35219</td>\n",
       "      <td>Escherichia coli (strain K12)</td>\n",
       "      <td>['P0A870', 'P30148']</td>\n",
       "      <td>TALB_ECOLI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3155</th>\n",
       "      <td>ytfP</td>\n",
       "      <td>98.0</td>\n",
       "      <td>0.9</td>\n",
       "      <td>12.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>conserved protein</td>\n",
       "      <td>Cytoplasmic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MRIFVYGSLRHKQGNSHWMTNAQLLGDFSIDNYQLYSLGHYPGAVP...</td>\n",
       "      <td>113</td>\n",
       "      <td>12866</td>\n",
       "      <td>Escherichia coli (strain K12)</td>\n",
       "      <td>['P0AE48', 'P39323', 'Q2M683']</td>\n",
       "      <td>YTFP_ECOLI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3156</th>\n",
       "      <td>yzfA</td>\n",
       "      <td>88.0</td>\n",
       "      <td>0.4</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>predicted protein</td>\n",
       "      <td>Cytoplasmic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MHMVTYPCLTSRRFQLALIHRRVVDKRTSMHSRTASESTGARIHRP...</td>\n",
       "      <td>89</td>\n",
       "      <td>10585</td>\n",
       "      <td>Escherichia coli (strain K12)</td>\n",
       "      <td>A0A385XK32</td>\n",
       "      <td>A0A385XK32_ECOLI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3157</th>\n",
       "      <td>chpB</td>\n",
       "      <td>94.0</td>\n",
       "      <td>0.4</td>\n",
       "      <td>5.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>toxin of the ChpB-ChpS toxin-antitoxin system</td>\n",
       "      <td>Cytoplasmic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MRITIKRWGNSAGMVIPNIVMKELNLQPGQSVEAQVSNNQLILTPI...</td>\n",
       "      <td>83</td>\n",
       "      <td>9272</td>\n",
       "      <td>Escherichia coli (strain K12)</td>\n",
       "      <td>['P08365', 'P76803', 'Q2M681']</td>\n",
       "      <td>CHPS_ECOLI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3158</th>\n",
       "      <td>nrdG</td>\n",
       "      <td>69.0</td>\n",
       "      <td>0.4</td>\n",
       "      <td>7.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>anaerobic ribonucleotide reductase activating ...</td>\n",
       "      <td>Cytoplasmic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MNYHQYYPVDIVNGPGTRCTLFVSGCVHECPGCYNKSTWRVNSGQP...</td>\n",
       "      <td>154</td>\n",
       "      <td>17446</td>\n",
       "      <td>Escherichia coli (strain K12)</td>\n",
       "      <td>['P0A9N8', 'P39329', 'Q2M670']</td>\n",
       "      <td>NRDG_ECOLI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3159</th>\n",
       "      <td>yjgF</td>\n",
       "      <td>76.0</td>\n",
       "      <td>1.1</td>\n",
       "      <td>15.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ketoacid-binding protein</td>\n",
       "      <td>Cytoplasmic</td>\n",
       "      <td>1QU9</td>\n",
       "      <td>55298; YjgF-like</td>\n",
       "      <td>MSKTIATENAPAAIGPYVQGVDLGNMIITSGQIPVNPKTGEVPADV...</td>\n",
       "      <td>128</td>\n",
       "      <td>13612</td>\n",
       "      <td>Escherichia coli (strain K12)</td>\n",
       "      <td>['P0AF93', 'P39330', 'P76806', 'Q2M664']</td>\n",
       "      <td>RIDA_ECOLI</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3160 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Gene Name  Solubility(%)  Yield(uM)  Yield(ug/ml)  Minus Sol  TF Sol  \\\n",
       "0         thrB           32.0        2.3          78.0        NaN     NaN   \n",
       "1         thrC           18.0        1.7          78.0      46.53   45.80   \n",
       "2         yaaX           78.0        1.2          14.0        NaN     NaN   \n",
       "3         yaaA            7.0        2.4          71.0      14.01   26.65   \n",
       "4         talB           85.0        2.7          94.0        NaN     NaN   \n",
       "...        ...            ...        ...           ...        ...     ...   \n",
       "3155      ytfP           98.0        0.9          12.0        NaN     NaN   \n",
       "3156      yzfA           88.0        0.4           4.0        NaN     NaN   \n",
       "3157      chpB           94.0        0.4           5.0        NaN     NaN   \n",
       "3158      nrdG           69.0        0.4           7.0        NaN     NaN   \n",
       "3159      yjgF           76.0        1.1          15.0        NaN     NaN   \n",
       "\n",
       "      GroE Sol  KJE Sol  Minus(uM)  TF(uM)  ...  \\\n",
       "0          NaN      NaN        NaN     NaN  ...   \n",
       "1        83.44    63.61      0.863   0.761  ...   \n",
       "2          NaN      NaN        NaN     NaN  ...   \n",
       "3        52.96    17.69      1.199   1.014  ...   \n",
       "4          NaN      NaN        NaN     NaN  ...   \n",
       "...        ...      ...        ...     ...  ...   \n",
       "3155       NaN      NaN        NaN     NaN  ...   \n",
       "3156       NaN      NaN        NaN     NaN  ...   \n",
       "3157       NaN      NaN        NaN     NaN  ...   \n",
       "3158       NaN      NaN        NaN     NaN  ...   \n",
       "3159       NaN      NaN        NaN     NaN  ...   \n",
       "\n",
       "                               Gene product description  Cell location  \\\n",
       "0                                     homoserine kinase    Cytoplasmic   \n",
       "1                                    threonine synthase    Cytoplasmic   \n",
       "2                                     predicted protein    Periplasmic   \n",
       "3                                     conserved protein    Cytoplasmic   \n",
       "4                                       transaldolase B    Cytoplasmic   \n",
       "...                                                 ...            ...   \n",
       "3155                                  conserved protein    Cytoplasmic   \n",
       "3156                                  predicted protein    Cytoplasmic   \n",
       "3157      toxin of the ChpB-ChpS toxin-antitoxin system    Cytoplasmic   \n",
       "3158  anaerobic ribonucleotide reductase activating ...    Cytoplasmic   \n",
       "3159                           ketoacid-binding protein    Cytoplasmic   \n",
       "\n",
       "      Structure (PDB) id                                    SCOP assignment  \\\n",
       "0                    NaN  54211; Ribosomal protein S5 domain 2-like 5506...   \n",
       "1                    NaN  53686; Tryptophan synthase beta subunit-like P...   \n",
       "2                    NaN                                                NaN   \n",
       "3                    NaN                                                NaN   \n",
       "4                   1ONR                                    51569; Aldolase   \n",
       "...                  ...                                                ...   \n",
       "3155                 NaN                                                NaN   \n",
       "3156                 NaN                                                NaN   \n",
       "3157                 NaN                                                NaN   \n",
       "3158                 NaN                                                NaN   \n",
       "3159                1QU9                                   55298; YjgF-like   \n",
       "\n",
       "                                               Sequence  Sequence length  \\\n",
       "0     MVKVYAPASSANMSVGFDVLGAAVTPVDGALLGDVVTVEAAETFSL...              310   \n",
       "1     MKLYNLKDHNEQVSFAQAVTQGLGKNQGLFFPHDLPEFSLTEIDEM...              428   \n",
       "2     MKKMQSIVLALSLVLVAPMAAQAAEITLVPSVKLQIGDRDNRGYYW...               98   \n",
       "3     MLILISPAKTLDYQSPLTTTRYTLPELLDNSQQLIHEARKLTPPQI...              258   \n",
       "4     MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...              317   \n",
       "...                                                 ...              ...   \n",
       "3155  MRIFVYGSLRHKQGNSHWMTNAQLLGDFSIDNYQLYSLGHYPGAVP...              113   \n",
       "3156  MHMVTYPCLTSRRFQLALIHRRVVDKRTSMHSRTASESTGARIHRP...               89   \n",
       "3157  MRITIKRWGNSAGMVIPNIVMKELNLQPGQSVEAQVSNNQLILTPI...               83   \n",
       "3158  MNYHQYYPVDIVNGPGTRCTLFVSGCVHECPGCYNKSTWRVNSGQP...              154   \n",
       "3159  MSKTIATENAPAAIGPYVQGVDLGNMIITSGQIPVNPKTGEVPADV...              128   \n",
       "\n",
       "      Sequence mass                               Organism  \\\n",
       "0             33624  Escherichia coli (strain K12 / DH10B)   \n",
       "1             47114          Escherichia coli (strain K12)   \n",
       "2             11355          Escherichia coli (strain K12)   \n",
       "3             29586  Escherichia coli (strain K12 / DH10B)   \n",
       "4             35219          Escherichia coli (strain K12)   \n",
       "...             ...                                    ...   \n",
       "3155          12866          Escherichia coli (strain K12)   \n",
       "3156          10585          Escherichia coli (strain K12)   \n",
       "3157           9272          Escherichia coli (strain K12)   \n",
       "3158          17446          Escherichia coli (strain K12)   \n",
       "3159          13612          Escherichia coli (strain K12)   \n",
       "\n",
       "                                         UP_ID           UP_NAME  \n",
       "0                                       B1XBC8        KHSE_ECODH  \n",
       "1                         ['P00934', 'Q6LEK8']        THRC_ECOLI  \n",
       "2                                       P75616        YAAX_ECOLI  \n",
       "3                                       B1XBD1        YAAA_ECODH  \n",
       "4                         ['P0A870', 'P30148']        TALB_ECOLI  \n",
       "...                                        ...               ...  \n",
       "3155            ['P0AE48', 'P39323', 'Q2M683']        YTFP_ECOLI  \n",
       "3156                                A0A385XK32  A0A385XK32_ECOLI  \n",
       "3157            ['P08365', 'P76803', 'Q2M681']        CHPS_ECOLI  \n",
       "3158            ['P0A9N8', 'P39329', 'Q2M670']        NRDG_ECOLI  \n",
       "3159  ['P0AF93', 'P39330', 'P76806', 'Q2M664']        RIDA_ECOLI  \n",
       "\n",
       "[3160 rows x 29 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# data = pd.read_csv('../Data/Sequencing/sequenced_data.csv')\n",
    "data = pd.read_csv('../Data/Processed Data/complete_data.csv')\n",
    "data = data.copy()\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "def toNumeric(data):\n",
    "    char = list(data.unique())\n",
    "    temp = []\n",
    "    for x in data:\n",
    "        temp.append(char.index(x))\n",
    "    return temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11])"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['Type of gene product numeric'] = toNumeric(data['Type of gene product'])\n",
    "data['Cell location numeric'] = toNumeric(data['Cell location'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "cell_id": "f703b118bc3345b88cc223438f1dd755",
    "deepnote_cell_height": 135,
    "deepnote_cell_type": "code",
    "deepnote_to_be_reexecuted": false,
    "execution_millis": 1,
    "execution_start": 1670939983043,
    "source_hash": "fa2f3e9a",
    "tags": []
   },
   "outputs": [],
   "source": [
    "X = data[['Yield(uM)','Yield(ug/ml)','Calculated MW(kDa)','Calculated pI', 'Sequence length', 'Sequence mass','Type of gene product numeric', 'Cell location numeric']]\n",
    "# X = data[['Yield(uM)','Yield(ug/ml)','Calculated MW(kDa)','Calculated pI', 'Sequence length', 'Sequence mass',]]\n",
    "# X = data[['Yield(uM)','Yield(ug/ml)','Calculated MW(kDa)','Calculated pI', 'Type of gene product Integer']]\n",
    "y = data['Solubility(%)']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "cell_id": "68ce673b91544ba2bd217256639f9dfd",
    "deepnote_cell_height": 94,
    "deepnote_cell_type": "code",
    "deepnote_to_be_reexecuted": false,
    "execution_millis": 2,
    "execution_start": 1670939983047,
    "source_hash": "b42d3c7e",
    "tags": []
   },
   "outputs": [],
   "source": [
    "def normalize(col_data):\n",
    "    return (col_data - col_data.min())/(col_data.max()-col_data.min())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cell_id": "00003-d7204aae-13ac-4491-bc90-7609e62221c6",
    "deepnote_cell_height": 76,
    "deepnote_cell_type": "code",
    "deepnote_to_be_reexecuted": false,
    "execution_millis": 4,
    "execution_start": 1670939983051,
    "source_hash": "84e17419"
   },
   "outputs": [],
   "source": [
    "X = data.iloc[:, 29:2029]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "cell_id": "e58c7f2fc7794d92bf66cc0135ddb3d0",
    "deepnote_cell_height": 112,
    "deepnote_cell_type": "code",
    "deepnote_to_be_reexecuted": false,
    "execution_millis": 3,
    "execution_start": 1670939983057,
    "source_hash": "b39d4967",
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/__/8qxqknx91_l7s7jy54cf7p2h0000gn/T/ipykernel_28695/3188529497.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  X[column] = normalize(X[column])\n"
     ]
    }
   ],
   "source": [
    "for column in X:\n",
    "    X[column] = normalize(X[column])\n",
    "X = round(X,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "cell_id": "00004-8e137585-5dc5-4dd1-a4f0-a6948a1c097a",
    "deepnote_cell_height": 76,
    "deepnote_cell_type": "code",
    "deepnote_to_be_reexecuted": false,
    "execution_millis": 1,
    "execution_start": 1670939983106,
    "source_hash": "dd4e21b0"
   },
   "outputs": [],
   "source": [
    "y = round(normalize(y),2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Yield(uM)</th>\n",
       "      <th>Yield(ug/ml)</th>\n",
       "      <th>Calculated MW(kDa)</th>\n",
       "      <th>Calculated pI</th>\n",
       "      <th>Sequence length</th>\n",
       "      <th>Sequence mass</th>\n",
       "      <th>Type of gene product numeric</th>\n",
       "      <th>Cell location numeric</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.15</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.26</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.11</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.31</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.31</td>\n",
       "      <td>0.31</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.08</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.76</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.15</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.53</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.17</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.23</td>\n",
       "      <td>0.23</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3155</th>\n",
       "      <td>0.06</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.47</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3156</th>\n",
       "      <td>0.03</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.85</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3157</th>\n",
       "      <td>0.03</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.30</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.15</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3158</th>\n",
       "      <td>0.03</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.43</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3159</th>\n",
       "      <td>0.07</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.38</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.40</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3160 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Yield(uM)  Yield(ug/ml)  Calculated MW(kDa)  Calculated pI  \\\n",
       "0          0.15          0.21                0.21           0.26   \n",
       "1          0.11          0.21                0.31           0.21   \n",
       "2          0.08          0.03                0.06           0.76   \n",
       "3          0.15          0.19                0.18           0.53   \n",
       "4          0.17          0.25                0.22           0.22   \n",
       "...         ...           ...                 ...            ...   \n",
       "3155       0.06          0.03                0.07           0.47   \n",
       "3156       0.03          0.01                0.05           0.85   \n",
       "3157       0.03          0.01                0.07           0.30   \n",
       "3158       0.03          0.01                0.10           0.43   \n",
       "3159       0.07          0.04                0.07           0.38   \n",
       "\n",
       "      Sequence length  Sequence mass  Type of gene product numeric  \\\n",
       "0                0.22           0.22                          0.00   \n",
       "1                0.31           0.31                          0.00   \n",
       "2                0.07           0.07                          0.05   \n",
       "3                0.18           0.19                          0.05   \n",
       "4                0.23           0.23                          0.00   \n",
       "...               ...            ...                           ...   \n",
       "3155             0.08           0.08                          0.05   \n",
       "3156             0.06           0.06                          0.05   \n",
       "3157             0.06           0.06                          0.15   \n",
       "3158             0.11           0.11                          0.00   \n",
       "3159             0.09           0.08                          0.40   \n",
       "\n",
       "      Cell location numeric  \n",
       "0                      0.00  \n",
       "1                      0.00  \n",
       "2                      0.09  \n",
       "3                      0.00  \n",
       "4                      0.00  \n",
       "...                     ...  \n",
       "3155                   0.00  \n",
       "3156                   0.00  \n",
       "3157                   0.00  \n",
       "3158                   0.00  \n",
       "3159                   0.00  \n",
       "\n",
       "[3160 rows x 8 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "cell_id": "00005-56cf79eb-dc1f-4c8f-a58a-c90f3a2baa60",
    "deepnote_cell_height": 76,
    "deepnote_cell_type": "code",
    "deepnote_to_be_reexecuted": false,
    "execution_millis": 4,
    "execution_start": 1670939983107,
    "source_hash": "26d9c02a"
   },
   "outputs": [],
   "source": [
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "cell_id": "fc4496efb9bd45fe955a8cd3daaf3f53",
    "deepnote_cell_height": 189.96875,
    "deepnote_cell_type": "code",
    "deepnote_to_be_reexecuted": false,
    "execution_millis": 27,
    "execution_start": 1670939983112,
    "source_hash": "4f3922ae",
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-01-10 14:45:46.125425: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n"
     ]
    }
   ],
   "source": [
    "model = tf.keras.Sequential()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cell_id": "a231915018464435ba31efd861264716",
    "deepnote_cell_height": 94,
    "deepnote_cell_type": "code",
    "deepnote_to_be_reexecuted": false,
    "execution_millis": 31,
    "execution_start": 1670939983151,
    "source_hash": "6c8bc719",
    "tags": []
   },
   "outputs": [],
   "source": [
    "model.add(layers.Dense(3, input_dim=6, activation='sigmoid'))\n",
    "model.add(layers.Dense(1, activation='relu'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cell_id": "00006-f7a89ffa-9195-465e-8a48-c338a1774add",
    "deepnote_cell_height": 1550.859375,
    "deepnote_cell_type": "code",
    "deepnote_output_heights": [
     null,
     605.5
    ],
    "deepnote_to_be_reexecuted": false,
    "execution_millis": 747,
    "execution_start": 1670939983200,
    "source_hash": "dec5dbf2"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 1/200\n",
      "WARNING:tensorflow:Model was constructed with shape (None, 6) for input KerasTensor(type_spec=TensorSpec(shape=(None, 6), dtype=tf.float32, name='dense_input'), name='dense_input', description=\"created by layer 'dense_input'\"), but it was called on an input with incompatible shape (None, 1).\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "in user code:\n\n    File \"/shared-libs/python3.9/py/lib/python3.9/site-packages/keras/engine/training.py\", line 1160, in train_function  *\n        return step_function(self, iterator)\n    File \"/shared-libs/python3.9/py/lib/python3.9/site-packages/keras/engine/training.py\", line 1146, in step_function  **\n        outputs = model.distribute_strategy.run(run_step, args=(data,))\n    File \"/shared-libs/python3.9/py/lib/python3.9/site-packages/keras/engine/training.py\", line 1135, in run_step  **\n        outputs = model.train_step(data)\n    File \"/shared-libs/python3.9/py/lib/python3.9/site-packages/keras/engine/training.py\", line 993, in train_step\n        y_pred = self(x, training=True)\n    File \"/shared-libs/python3.9/py/lib/python3.9/site-packages/keras/utils/traceback_utils.py\", line 70, in error_handler\n        raise e.with_traceback(filtered_tb) from None\n    File \"/shared-libs/python3.9/py/lib/python3.9/site-packages/keras/engine/input_spec.py\", line 277, in assert_input_compatibility\n        raise ValueError(\n\n    ValueError: Exception encountered when calling layer \"sequential\" \"                 f\"(type Sequential).\n    \n    Input 0 of layer \"dense\" is incompatible with the layer: expected axis -1 of input shape to have value 6, but received input with shape (None, 1)\n    \n    Call arguments received by layer \"sequential\" \"                 f\"(type Sequential):\n      • inputs=tf.Tensor(shape=(None, 1), dtype=float64)\n      • training=True\n      • mask=None\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [12], line 7\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# compile the keras model\u001b[39;00m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;66;03m#model.compile(loss='mean_absolute_error', optimizer=tf.keras.optimizers.Adam(learning_rate=0.0001),\u001b[39;00m\n\u001b[1;32m      3\u001b[0m      \u001b[38;5;66;03m#metrics=[tf.keras.metrics.MeanSquaredError(name='MSE'), tf.keras.metrics.RootMeanSquaredError(name='MRSE')])\u001b[39;00m\n\u001b[1;32m      5\u001b[0m model\u001b[38;5;241m.\u001b[39mcompile(loss\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmean_absolute_error\u001b[39m\u001b[38;5;124m'\u001b[39m, optimizer\u001b[38;5;241m=\u001b[39mtf\u001b[38;5;241m.\u001b[39mkeras\u001b[38;5;241m.\u001b[39moptimizers\u001b[38;5;241m.\u001b[39mAdam(\u001b[38;5;241m0.001\u001b[39m),  metrics\u001b[38;5;241m=\u001b[39m[tf\u001b[38;5;241m.\u001b[39mkeras\u001b[38;5;241m.\u001b[39mmetrics\u001b[38;5;241m.\u001b[39mMeanSquaredError(name\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mMSE\u001b[39m\u001b[38;5;124m'\u001b[39m), tf\u001b[38;5;241m.\u001b[39mkeras\u001b[38;5;241m.\u001b[39mmetrics\u001b[38;5;241m.\u001b[39mRootMeanSquaredError(name\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mMRSE\u001b[39m\u001b[38;5;124m'\u001b[39m)])\n\u001b[0;32m----> 7\u001b[0m history \u001b[38;5;241m=\u001b[39m \u001b[43mmodel\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfit\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m      8\u001b[0m \u001b[43m    \u001b[49m\u001b[43mX_train\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m      9\u001b[0m \u001b[43m    \u001b[49m\u001b[43my_train\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     10\u001b[0m \u001b[43m    \u001b[49m\u001b[43mbatch_size\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m20\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     11\u001b[0m \u001b[43m    \u001b[49m\u001b[43mepochs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m200\u001b[39;49m\n\u001b[1;32m     12\u001b[0m \u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/shared-libs/python3.9/py/lib/python3.9/site-packages/keras/utils/traceback_utils.py:70\u001b[0m, in \u001b[0;36mfilter_traceback.<locals>.error_handler\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     67\u001b[0m     filtered_tb \u001b[38;5;241m=\u001b[39m _process_traceback_frames(e\u001b[38;5;241m.\u001b[39m__traceback__)\n\u001b[1;32m     68\u001b[0m     \u001b[38;5;66;03m# To get the full stack trace, call:\u001b[39;00m\n\u001b[1;32m     69\u001b[0m     \u001b[38;5;66;03m# `tf.debugging.disable_traceback_filtering()`\u001b[39;00m\n\u001b[0;32m---> 70\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m e\u001b[38;5;241m.\u001b[39mwith_traceback(filtered_tb) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;28mNone\u001b[39m\n\u001b[1;32m     71\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[1;32m     72\u001b[0m     \u001b[38;5;28;01mdel\u001b[39;00m filtered_tb\n",
      "File \u001b[0;32m/tmp/__autograph_generated_filetuws2_vb.py:15\u001b[0m, in \u001b[0;36mouter_factory.<locals>.inner_factory.<locals>.tf__train_function\u001b[0;34m(iterator)\u001b[0m\n\u001b[1;32m     13\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m     14\u001b[0m     do_return \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[0;32m---> 15\u001b[0m     retval_ \u001b[38;5;241m=\u001b[39m ag__\u001b[38;5;241m.\u001b[39mconverted_call(ag__\u001b[38;5;241m.\u001b[39mld(step_function), (ag__\u001b[38;5;241m.\u001b[39mld(\u001b[38;5;28mself\u001b[39m), ag__\u001b[38;5;241m.\u001b[39mld(iterator)), \u001b[38;5;28;01mNone\u001b[39;00m, fscope)\n\u001b[1;32m     16\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m:\n\u001b[1;32m     17\u001b[0m     do_return \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m\n",
      "\u001b[0;31mValueError\u001b[0m: in user code:\n\n    File \"/shared-libs/python3.9/py/lib/python3.9/site-packages/keras/engine/training.py\", line 1160, in train_function  *\n        return step_function(self, iterator)\n    File \"/shared-libs/python3.9/py/lib/python3.9/site-packages/keras/engine/training.py\", line 1146, in step_function  **\n        outputs = model.distribute_strategy.run(run_step, args=(data,))\n    File \"/shared-libs/python3.9/py/lib/python3.9/site-packages/keras/engine/training.py\", line 1135, in run_step  **\n        outputs = model.train_step(data)\n    File \"/shared-libs/python3.9/py/lib/python3.9/site-packages/keras/engine/training.py\", line 993, in train_step\n        y_pred = self(x, training=True)\n    File \"/shared-libs/python3.9/py/lib/python3.9/site-packages/keras/utils/traceback_utils.py\", line 70, in error_handler\n        raise e.with_traceback(filtered_tb) from None\n    File \"/shared-libs/python3.9/py/lib/python3.9/site-packages/keras/engine/input_spec.py\", line 277, in assert_input_compatibility\n        raise ValueError(\n\n    ValueError: Exception encountered when calling layer \"sequential\" \"                 f\"(type Sequential).\n    \n    Input 0 of layer \"dense\" is incompatible with the layer: expected axis -1 of input shape to have value 6, but received input with shape (None, 1)\n    \n    Call arguments received by layer \"sequential\" \"                 f\"(type Sequential):\n      • inputs=tf.Tensor(shape=(None, 1), dtype=float64)\n      • training=True\n      • mask=None\n"
     ]
    }
   ],
   "source": [
    "# compile the keras model\n",
    "#model.compile(loss='mean_absolute_error', optimizer=tf.keras.optimizers.Adam(learning_rate=0.0001),\n",
    "     #metrics=[tf.keras.metrics.MeanSquaredError(name='MSE'), tf.keras.metrics.RootMeanSquaredError(name='MRSE')])\n",
    "\n",
    "model.compile(loss='mean_absolute_error', optimizer=tf.keras.optimizers.Adam(0.001),  metrics=[tf.keras.metrics.MeanSquaredError(name='MSE'), tf.keras.metrics.RootMeanSquaredError(name='MRSE')])\n",
    "\n",
    "history = model.fit(\n",
    "    X_train,\n",
    "    y_train,\n",
    "    batch_size=20,\n",
    "    epochs=200\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cell_id": "53c28e1d7d8c45c890c1bb26203e5c24",
    "deepnote_cell_height": 106.6875,
    "deepnote_cell_type": "code",
    "deepnote_to_be_reexecuted": true,
    "execution_millis": 718,
    "execution_start": 1670416312100,
    "source_hash": "1bb4d685",
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO:tensorflow:Assets written to: ../models/models_SNN/assets\n"
     ]
    }
   ],
   "source": [
    "model.save('../models/models_SNN')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cell_id": "00007-1e2dd1f2-58a1-4c18-9414-e635287bd74c",
    "deepnote_cell_height": 106.6875,
    "deepnote_cell_type": "code",
    "deepnote_to_be_reexecuted": true,
    "execution_millis": 231,
    "execution_start": 1670411335113,
    "source_hash": "43052034"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20/20 [==============================] - 0s 1ms/step - loss: 0.6184 - MSE: 0.0426 - MRSE: 0.2065\n"
     ]
    }
   ],
   "source": [
    "eval_results = model.evaluate(X_test, y_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cell_id": "00008-5abbd495-eeb8-45ba-8d99-5b1eb6f5e269",
    "deepnote_cell_height": 106.6875,
    "deepnote_cell_type": "code",
    "deepnote_to_be_reexecuted": true,
    "execution_millis": 4,
    "execution_start": 1670411342838,
    "source_hash": "1004deae"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loss, accuracy [0.6184118986129761, 0.04262311011552811, 0.20645365118980408]\n"
     ]
    }
   ],
   "source": [
    "print(\"loss, accuracy\", eval_results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cell_id": "d37cc131e949496b847a00373082eba4",
    "deepnote_cell_height": 106.6875,
    "deepnote_cell_type": "code",
    "deepnote_to_be_reexecuted": true,
    "execution_millis": 3,
    "execution_start": 1668671875605,
    "source_hash": "546c4539",
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loss, accuracy 4.173898324370384 %\n"
     ]
    }
   ],
   "source": [
    "print(\"loss, accuracy\", eval_results * 100, '%')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "created_in_deepnote_cell": true,
    "deepnote_cell_type": "markdown",
    "tags": []
   },
   "source": [
    "<a style='text-decoration:none;line-height:16px;display:flex;color:#5B5B62;padding:10px;justify-content:end;' href='https://deepnote.com?utm_source=created-in-deepnote-cell&projectId=8ea8948f-b978-492f-a6dd-8ac5feb4b472' target=\"_blank\">\n",
    "<img alt='Created in deepnote.com' style='display:inline;max-height:16px;margin:0px;margin-right:7.5px;' src='data:image/svg+xml;base64,PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0iVVRGLTgiPz4KPHN2ZyB3aWR0aD0iODBweCIgaGVpZ2h0PSI4MHB4IiB2aWV3Qm94PSIwIDAgODAgODAiIHZlcnNpb249IjEuMSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB4bWxuczp4bGluaz0iaHR0cDovL3d3dy53My5vcmcvMTk5OS94bGluayI+CiAgICA8IS0tIEdlbmVyYXRvcjogU2tldGNoIDU0LjEgKDc2NDkwKSAtIGh0dHBzOi8vc2tldGNoYXBwLmNvbSAtLT4KICAgIDx0aXRsZT5Hcm91cCAzPC90aXRsZT4KICAgIDxkZXNjPkNyZWF0ZWQgd2l0aCBTa2V0Y2guPC9kZXNjPgogICAgPGcgaWQ9IkxhbmRpbmciIHN0cm9rZT0ibm9uZSIgc3Ryb2tlLXdpZHRoPSIxIiBmaWxsPSJub25lIiBmaWxsLXJ1bGU9ImV2ZW5vZGQiPgogICAgICAgIDxnIGlkPSJBcnRib2FyZCIgdHJhbnNmb3JtPSJ0cmFuc2xhdGUoLTEyMzUuMDAwMDAwLCAtNzkuMDAwMDAwKSI+CiAgICAgICAgICAgIDxnIGlkPSJHcm91cC0zIiB0cmFuc2Zvcm09InRyYW5zbGF0ZSgxMjM1LjAwMDAwMCwgNzkuMDAwMDAwKSI+CiAgICAgICAgICAgICAgICA8cG9seWdvbiBpZD0iUGF0aC0yMCIgZmlsbD0iIzAyNjVCNCIgcG9pbnRzPSIyLjM3NjIzNzYyIDgwIDM4LjA0NzY2NjcgODAgNTcuODIxNzgyMiA3My44MDU3NTkyIDU3LjgyMTc4MjIgMzIuNzU5MjczOSAzOS4xNDAyMjc4IDMxLjY4MzE2ODMiPjwvcG9seWdvbj4KICAgICAgICAgICAgICAgIDxwYXRoIGQ9Ik0zNS4wMDc3MTgsODAgQzQyLjkwNjIwMDcsNzYuNDU0OTM1OCA0Ny41NjQ5MTY3LDcxLjU0MjI2NzEgNDguOTgzODY2LDY1LjI2MTk5MzkgQzUxLjExMjI4OTksNTUuODQxNTg0MiA0MS42NzcxNzk1LDQ5LjIxMjIyODQgMjUuNjIzOTg0Niw0OS4yMTIyMjg0IEMyNS40ODQ5Mjg5LDQ5LjEyNjg0NDggMjkuODI2MTI5Niw0My4yODM4MjQ4IDM4LjY0NzU4NjksMzEuNjgzMTY4MyBMNzIuODcxMjg3MSwzMi41NTQ0MjUgTDY1LjI4MDk3Myw2Ny42NzYzNDIxIEw1MS4xMTIyODk5LDc3LjM3NjE0NCBMMzUuMDA3NzE4LDgwIFoiIGlkPSJQYXRoLTIyIiBmaWxsPSIjMDAyODY4Ij48L3BhdGg+CiAgICAgICAgICAgICAgICA8cGF0aCBkPSJNMCwzNy43MzA0NDA1IEwyNy4xMTQ1MzcsMC4yNTcxMTE0MzYgQzYyLjM3MTUxMjMsLTEuOTkwNzE3MDEgODAsMTAuNTAwMzkyNyA4MCwzNy43MzA0NDA1IEM4MCw2NC45NjA0ODgyIDY0Ljc3NjUwMzgsNzkuMDUwMzQxNCAzNC4zMjk1MTEzLDgwIEM0Ny4wNTUzNDg5LDc3LjU2NzA4MDggNTMuNDE4MjY3Nyw3MC4zMTM2MTAzIDUzLjQxODI2NzcsNTguMjM5NTg4NSBDNTMuNDE4MjY3Nyw0MC4xMjg1NTU3IDM2LjMwMzk1NDQsMzcuNzMwNDQwNSAyNS4yMjc0MTcsMzcuNzMwNDQwNSBDMTcuODQzMDU4NiwzNy43MzA0NDA1IDkuNDMzOTE5NjYsMzcuNzMwNDQwNSAwLDM3LjczMDQ0MDUgWiIgaWQ9IlBhdGgtMTkiIGZpbGw9IiMzNzkzRUYiPjwvcGF0aD4KICAgICAgICAgICAgPC9nPgogICAgICAgIDwvZz4KICAgIDwvZz4KPC9zdmc+' > </img>\n",
    "Created in <span style='font-weight:600;margin-left:4px;'>Deepnote</span></a>"
   ]
  }
 ],
 "metadata": {
  "deepnote": {},
  "deepnote_execution_queue": [],
  "deepnote_notebook_id": "442e4f49-98b6-4548-aab2-a38a608686a7",
  "deepnote_persisted_session": {
   "createdAt": "2022-12-13T14:18:15.993Z"
  },
  "kernelspec": {
   "display_name": "Python 3.9.0 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  },
  "vscode": {
   "interpreter": {
    "hash": "aee8b7b246df8f9039afb4144a1f6fd8d2ca17a180786b69acc140d282b71a49"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
